### PR TITLE
RFC: Btree query API

### PR DESCRIPTION
**Pull request by Gankra**
This is an implementation of the core functionality of https://github.com/rust-lang/rfcs/pull/1195 which is not yet accepted (largely due to bikeshedding naming -- but also on what removal/entry APIs to expose).

Note that the vast majority of the code was generated by a program (see comments).

I believe I've exhaustively tested every single code path with my unit test.

Note that removal/entry APIs are not yet provided; I'm of the opinion that a major refactor of the internals to use parent pointers should be done first. Then search/removal/entry/range can basically share the exact same source code with no overhead. It's also expected to be substantially more efficient due to not needing to maintain an explicit search stack for insert/remove/range/entry.

### CODE REVIEW

**Comment by Gankro** [comment]
cc @apasel422 @gereeter

**Comment by apasel422** [comment]
@Gankro Wow, this is really cool, but definitely highlights the need for some kind of parametric mutability.

File: src/libcollections/btree/set.rs [hunk 1]
```diff
@@ -324,6 +324,40 @@ impl<T: Ord> BTreeSet<T> {
...
+    pub fn min(&self) -> Option<&T> {
+        self.map.min().map(|r| r.0)
+    }
+
+    /// Gets the maximum value in the set.
+    pub fn max(&self) -> Option<&T> {
+        self.map.max().map(|r| r.0)
+    }
+
+    /// Gets the closest value to the given key that is less than it.
```
Comment (apasel422) [inline src/libcollections/btree/set.rs]: s/key/value/

(Same for the following methods.)

File: src/libcollections/btree/node.rs [hunk 2]
```diff
@@ -887,14 +946,25 @@ impl<'a, K: 'a, V: 'a, NodeType> Handle<&'a mut Node<K, V>, handle::KV, NodeType
...
+    /// lifetime than `&self`.
+    pub fn into_left_edge_mut(self) -> Handle<&'a mut Node<K, V>, handle::Edge, NodeType> {
         Handle {
             node: &mut *self.node,
             index: self.index,
             marker: PhantomData,
         }
     }
+
+    /// Convert this handle into one pointing at the edge immediately to the left of the key/value
```
Comment (apasel422) [inline src/libcollections/btree/node.rs]: s/Convert/Converts/

File: src/libcollections/btree/node.rs [hunk 3]
```diff
@@ -887,14 +946,25 @@ impl<'a, K: 'a, V: 'a, NodeType> Handle<&'a mut Node<K, V>, handle::KV, NodeType
...
+    /// pair pointed-to by this handle. This is useful because it returns a reference with larger
```
Comment (apasel422) [inline src/libcollections/btree/node.rs]: s/larger/a larger/

File: src/libcollections/btree/set.rs (hunk 1 above)
Comment (Gankro) [inline src/libcollections/btree/set.rs]: All should presumably be element.

File: src/libcollections/btree/node.rs [hunk 4]
```diff
@@ -869,6 +906,28 @@ impl<'a, K: 'a, V: 'a, NodeType> Handle<&'a Node<K, V>, handle::KV, NodeType> {
             )
         }
     }
+
+    /// Convert this handle into one pointing at the edge immediately to the left of the key/value
```
Comment (apasel422) [inline src/libcollections/btree/node.rs]: See note for `into_right_edge_mut`.

File: src/libcollections/btree/node.rs [hunk 5]
```diff
@@ -869,6 +906,28 @@ impl<'a, K: 'a, V: 'a, NodeType> Handle<&'a Node<K, V>, handle::KV, NodeType> {
...
+    /// lifetime than `&self`.
+    pub fn into_left_edge(self) -> Handle<&'a Node<K, V>, handle::Edge, NodeType> {
+        Handle {
+            node: &*self.node,
+            index: self.index,
+            marker: PhantomData,
+        }
+    }
+
+    /// Convert this handle into one pointing at the edge immediately to the left of the key/value
```
Comment (apasel422) [inline src/libcollections/btree/node.rs]: See note for `into_right_edge_mut`.

**Comment by Gankro** [comment]
I'm just *really* happy I didn't have to write 12 node search procedures as well. I was dreading the impl until I realized Found/GoDown was basically sufficient.

File: src/libcollections/btree/map.rs [hunk 6]
```diff
@@ -228,6 +228,660 @@ impl<K: Ord, V> BTreeMap<K, V> {
...
+                            }})
+                        }} else {{
+                            unsafe {{
+                                let (k, v) = next.{dir}_kv{muta_ext}().into_kv{muta_ext}();
+                                mem::transmute(Some((&*k, v)))
+                            }}
+                        }};
+                    }},
+                    Internal(internal_handle) => {{
+                        temp_node = internal_handle.into_{dir}_edge{muta_ext}().into_edge{muta_ext}();
```
Comment (apasel422) [inline src/libcollections/btree/map.rs]: There was a tidy error on this line.

File: src/libcollections/btree/set.rs [hunk 7]
```diff
@@ -324,6 +324,40 @@ impl<T: Ord> BTreeSet<T> {
...
+        self.map.min().map(|r| r.0)
+    }
+
+    /// Gets the maximum element in the set.
+    pub fn max(&self) -> Option<&T> {
+        self.map.max().map(|r| r.0)
+    }
+
+    /// Gets the closest element to the given element that is less than it.
+    pub fn get_lt<Q: ?Sized>(&self, value: &Q) -> Option<&T>
```
Comment (apasel422) [inline src/libcollections/btree/set.rs]: Presumably these parameters should be renamed to `element`.

File: src/libcollectionstest/btree/map.rs [hunk 8]
```diff
@@ -76,6 +76,81 @@ fn test_basic_small() {
...
+    let min = gap;
+    let max = (size - 1) * gap;
+
+    let mut map = BTreeMap::new();
+
+    assert_eq!(map.min(), None);
+    assert_eq!(map.max(), None);
+    assert_eq!(map.get_lt(&0), None);
+    assert_eq!(map.get_gt(&0), None);
+    assert_eq!(map.get_leq(&0), None);
```
Comment (apasel422) [inline src/libcollectionstest/btree/map.rs]: s/leq/le/ (throughout)

**Comment by Gankro** [comment]
Bench results:

```
get_rand_10000_eq                       ... bench:          50 ns/iter (+/- 7)
get_rand_10000_ge                       ... bench:          53 ns/iter (+/- 2)
get_rand_10000_ge_range                 ... bench:         117 ns/iter (+/- 7)
get_rand_10000_gt                       ... bench:          56 ns/iter (+/- 6)
get_rand_10000_gt_range                 ... bench:         112 ns/iter (+/- 6)
get_rand_10000_le                       ... bench:          53 ns/iter (+/- 4)
get_rand_10000_le_range                 ... bench:         110 ns/iter (+/- 7)
get_rand_10000_lt                       ... bench:          58 ns/iter (+/- 5)
get_rand_10000_lt_range                 ... bench:         115 ns/iter (+/- 11)

get_rand_100_eq                       ... bench:          10 ns/iter (+/- 2)
get_rand_100_ge                         ... bench:          11 ns/iter (+/- 2)
get_rand_100_ge_range                   ... bench:          60 ns/iter (+/- 8)
get_rand_100_gt                         ... bench:          12 ns/iter (+/- 2)
get_rand_100_gt_range                   ... bench:          58 ns/iter (+/- 15)
get_rand_100_le                         ... bench:          11 ns/iter (+/- 2)
get_rand_100_le_range                   ... bench:          57 ns/iter (+/- 2)
get_rand_100_lt                         ... bench:          14 ns/iter (+/- 2)
get_rand_100_lt_range                   ... bench:          63 ns/iter (+/- 20)

get_seq_10000_eq                      ... bench:          35 ns/iter (+/- 5)
get_seq_10000_ge                        ... bench:          40 ns/iter (+/- 5)
get_seq_10000_ge_range                  ... bench:          93 ns/iter (+/- 7)
get_seq_10000_gt                        ... bench:          40 ns/iter (+/- 2)
get_seq_10000_gt_range                  ... bench:          98 ns/iter (+/- 5)
get_seq_10000_le                        ... bench:          35 ns/iter (+/- 2)
get_seq_10000_le_range                  ... bench:          97 ns/iter (+/- 25)
get_seq_10000_lt                        ... bench:          41 ns/iter (+/- 3)
get_seq_10000_lt_range                  ... bench:         103 ns/iter (+/- 9)

get_seq_100_eq                          ... bench:          11 ns/iter (+/- 1)
get_seq_100_ge                          ... bench:          12 ns/iter (+/- 0)
get_seq_100_ge_range                    ... bench:          64 ns/iter (+/- 9)
get_seq_100_gt                          ... bench:          13 ns/iter (+/- 3)
get_seq_100_gt_range                    ... bench:          64 ns/iter (+/- 8)
get_seq_100_le                          ... bench:          11 ns/iter (+/- 2)
get_seq_100_le_range                    ... bench:          63 ns/iter (+/- 153)
get_seq_100_lt                          ... bench:          15 ns/iter (+/- 1)
get_seq_100_lt_range                    ... bench:          68 ns/iter (+/- 18)

insert_rand_100                         ... bench:         105 ns/iter (+/- 11)
insert_rand_10000                       ... bench:         138 ns/iter (+/- 14)
insert_seq_100                          ... bench:         107 ns/iter (+/- 13)
insert_seq_10000                        ... bench:         201 ns/iter (+/- 36)

iter_100000_plain                       ... bench:     891,492 ns/iter (+/- 93,913)
iter_100000_with_queries_backward       ... bench:   4,648,407 ns/iter (+/- 690,988)
iter_100000_with_queries_forward        ... bench:   4,580,392 ns/iter (+/- 760,890)

iter_1000_plain                         ... bench:       8,500 ns/iter (+/- 1,250)
iter_1000_with_queries_backward         ... bench:      31,266 ns/iter (+/- 2,909)
iter_1000_with_queries_forward          ... bench:      29,045 ns/iter (+/- 2,350)

iter_20_plain                           ... bench:         170 ns/iter (+/- 19)
iter_20_with_queries_backward           ... bench:         207 ns/iter (+/- 35)
iter_20_with_queries_forward            ... bench:         159 ns/iter (+/- 38)
```

File: src/libcollections/btree/map.rs [hunk 9]
```diff
@@ -228,6 +228,665 @@ impl<K: Ord, V> BTreeMap<K, V> {
         }
     }
 
+// Rust's macro system is so attrocious at being generic over mutability
```
Comment (apasel422) [inline src/libcollections/btree/map.rs]: s/attrocious/atrocious/

Comment (Gankro) [inline src/libcollections/btree/map.rs]: This correction made my day

**Comment by Gankro** [comment]
RFC closed.

**Comment by Gankro** [comment]
@apasel422 Note that you might want to steal these bench changes for profiling parent pointer stuff.

**Pull request closed without merging**